# Bài tập 6

Nguyễn Trọng Khiêm - 1712529

## Câu 1 (1 điểm)

Việc sử dụng một tập con $\mathcal{H}' \subset \mathcal{H}$ sẽ làm tăng deterministic noise, vì $\mathcal{H}'$ là con của $\mathcal{H}$ nên $\mathcal{H}'$ sẽ không đủ mạnh để xấp xỉ $f$ (hay ít nhất thì nó không đủ mạnh như $\mathcal{H}$).

Vậy em chọn đáp án [b] In general, deterministic noise will increase.

## Câu 2 (1 điểm)

In [1]:
import numpy as np

# Load data
data_train = np.loadtxt("in.dta")
X_train, y_train = data_train[:,:2], data_train[:,2]

data_test = np.loadtxt("out.dta")
X_test, y_test = data_test[:,:2], data_test[:,2]

# Transform data
def phi(X):
    Z = np.array([np.ones(X.shape[0]), 
                  X[:,0], 
                  X[:,1],
                  X[:,0]**2, 
                  X[:,1]**2, 
                  X[:,0] * X[:,1],
                  np.abs(X[:,0] - X[:,1]), 
                  np.abs(X[:,0] + X[:,1])]).T
    return Z

Z_train = phi(X_train)

# Fit the data
Z = Z_train

Z_dagger = np.dot(np.linalg.inv(np.dot(Z.T, Z)), Z.T)
w_tilde = np.dot(Z_dagger, y_train)

def predict(X):
    Z = phi(X)
    return np.sign(Z @ w_tilde)

# Tính E_in, E_out
E_in = np.mean(predict(X_train) != y_train)
print("E_in =", E_in)

E_out = np.mean(predict(X_test) != y_test)
print("E_out =", E_out)

E_in = 0.02857142857142857
E_out = 0.084


Vậy em chọn đáp án [a] 0.03, 0.08

## Câu 3 (1 điểm)

In [2]:
Z_train = phi(X_train)

for k in range(-5, 5):
    # Fit the data
    Z = Z_train
    
    Z_dagger_reg = np.dot(np.linalg.inv(np.dot(Z.T, Z) + 10**k * np.identity(Z.shape[1])), Z.T)
    w_tilde_reg = np.dot(Z_dagger_reg, y_train)
    def predict(X):
        Z = phi(X)
        return np.sign(Z @ w_tilde_reg)
    print("k =", k)
    E_in = np.mean(predict(X_train) != y_train)
    print("E_in =", E_in)
    E_out = np.mean(predict(X_test) != y_test)
    print("E_out =", E_out)
    print('\n')

k = -5
E_in = 0.02857142857142857
E_out = 0.084


k = -4
E_in = 0.02857142857142857
E_out = 0.084


k = -3
E_in = 0.02857142857142857
E_out = 0.08


k = -2
E_in = 0.02857142857142857
E_out = 0.084


k = -1
E_in = 0.02857142857142857
E_out = 0.056


k = 0
E_in = 0.0
E_out = 0.092


k = 1
E_in = 0.05714285714285714
E_out = 0.124


k = 2
E_in = 0.2
E_out = 0.228


k = 3
E_in = 0.37142857142857144
E_out = 0.436


k = 4
E_in = 0.42857142857142855
E_out = 0.452




Với $k=-3$, $E_{in} \approx 0.03, E_{out} = 0.08$.  
Vậy em chọn đáp án [d] 0.03, 0.08

## Câu 4 (1 điểm)

Với $k=3$, $E_{in} \approx 0.4, E_{out} \approx 0.4$.  
Vậy em chọn đáp án [e] 0.4, 0.4

## Câu 5 (1 điểm)

Theo như trên, $E_{out}$ nhỏ nhất khi $k=-1$.
Vậy em chọn đáp án [d] -1

## Câu 6 (1 điểm)

$min E_{out} = 0.056 \approx 0.06$.
Vậy em chọn đáp án [b] 0.06

## Câu 7 (1 điểm)

Theo đề bài, $\mathcal {H}(Q,C,Q_0)$ nghĩa là ta gán tất cả các hệ số $w_q$ với $q \geq Q_0$ thành $C$.  
Khi đó, $\mathcal {H}(10,0,Q_0)$ nghĩa là ta gán tất cả các hệ số $w_q$ với $q \geq Q_0$ thành $0$. 
Cụ thể, $\mathcal {H}(10,0,3)$ sẽ bao gồm tất cả các đa thức bậc hai và $\mathcal {H}(10,0,4)$ sẽ gồm các đa thức bậc ba. Khi đó giao của hai tập này sẽ là $\mathcal {H}_2$.

Vậy em chọn đáp án [c]

## Câu 8 (1 điểm)

Một lần backpropagation với một mẫu dữ liệu $(\mathbf{x}, y)$ mà đề bài hỏi gồm các bước:
1. Bắt đầu từ véc-tơ input $\mathbf{x}$, thực hiện lan truyền tiến qua các tầng của mạng để tính ra các giá trị đầu ra $a$ của các nơ-ron. Lưu ý: ở đây mình dùng ký hiệu $a^{(l)}_i$ để chỉ giá trị đầu ra của nơ-ron $i$ ở tầng $l$, còn trong đề bài là dùng ký hiệu $x^{(l)}_i$. Số phép tính $w^{(l)}_{ij}a^{(l-1)}_i$ ở bước lan truyền tiến này là: $3 \times 6 + 1 \times 4 = 22$
2. Từ $\delta$ ở tầng cuối, thực hiện lan truyền ngược để tính $\delta$ của các nơ-ron ở các tầng trước đó.Ta sẽ không cần tính $\delta$ của các nơ-ron output cũng như của nơ-ron bias ở lớp 1 .Số phép tính $w^{(l)}_{ij}\delta^{(l)}_j$ ở bước này sẽ là: 3
3. Với mỗi trọng số $w^{(l)}_{ij}$, tính $\frac{\partial e}{\partial w^{(l)}_{ij}}$ bằng cách lấy $a$ ở một đầu (nơ-ron $i$ ở tầng $l-1$) nhân với $\delta$ ở đầu kia (nơ-ron $j$ ở tầng $l$). Số phép tính $a^{(l-1)}_i\delta^{(l)}_j$ ở bước nàylà: $3 \times 6 + 1 \times 4 = 22$    
Vậy tổng số phép tính phải thực hiện là $22+3+22=47$, gần với đáp án [d] 45.

Vậy em chọn đáp án [d] 45

## Câu 9 (1 điểm)

Với một mạng NNs gồm 10 input units, 1 output unit, 36 hidden units, số trọng số (hay số cạnh) tối thiểu đạt được khi mạng có kiến trúc gồm 18 hidden layers, mỗi hidden layer gồm 1 unit thường và 1 bias unit:  
$d = [9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]$, với $d^{(l)}$ là số lượng units ở tầng $l$ mà chưa tính bias unit.  
Khi đó tổng số cạnh của mạng là: $10 \times 1 + 17 \times 2 + 2 \times 1 = 46$

Vậy em chọn đáp án [a] 46

## Câu 10 (1 điểm)

Với một mạng NNs gồm 10 input units, 1 output unit, 36 hidden units, số trọng số (hay số cạnh) tối thiểu đạt được khi mạng có kiến trúc gồm 2 hidden layers, layer 1 gồm 22 units, layer 2 gồm 14 units (bao gồm cả bias unit):  
$d = [9, 21, 13, 1]$, với $d^{(l)}$ là số lượng units ở tầng $l$ mà chưa tính bias unit.  
Khi đó tổng số cạnh của mạng là: $10 \times 21 + 22 \times 13 + 14 \times 1 = 510$

Vậy em chọn đáp án [e] 510